src/main/java/sg/ncl/MainController.java

```diff
@@ -471,10 +481,106 @@ public String loginSubmit(
+        HttpEntity<String> request = new HttpEntity<>(obj.toString(), headers);
+        restTemplate.setErrorHandler(new MyResponseErrorHandler());
+        ResponseEntity response = null;
+        try {
+            response = restTemplate.exchange(properties.getPasswordResetRequestURI(), HttpMethod.POST, request, String.class);
+        } catch (RestClientException e) {
+            log.warn("Cannot connect to sio for password reset email: {}", e);
+            passwordResetRequestForm.setErrMsg("Cannot connect. Server may be down!");
+            return FORGET_PSWD_PAGE;
+        }
+
+        if (RestUtil.isError(response.getStatusCode())) {
+            log.warn("Server responded error for password reset email: {}", response.getStatusCode());
+            passwordResetRequestForm.setErrMsg("Email not registered. Please use a different email address.");
+            return FORGET_PSWD_PAGE;
+        }
+
+        log.info("Password reset email sent for {}", email);
+        return "password_reset_email_sent";
+    }
+
+    // triggered when user clicks password reset link in the email
+    @RequestMapping(path = "/passwordReset", params = {"key"})
+    public String passwordResetNewPassword(@NotNull @RequestParam("key") final String key, Model model, HttpSession session) {
+        model.addAttribute("passwordResetForm", new PasswordResetForm());
+        session.setAttribute("key", key);
+        // redirect to the page for user to enter new password
+        return FORGET_PSWD_NEW_PSWD_PAGE;
+    }
+
+    // actual call to sio to reset password
+    @RequestMapping(path = "/password_reset")
+    public String resetPassword(@ModelAttribute("passwordResetForm") PasswordResetForm passwordResetForm,
+                                HttpSession session) throws IOException
+    {
+        if(!passwordResetForm.isPasswordOk()) {
+            return FORGET_PSWD_NEW_PSWD_PAGE;
+        }
+
+        JSONObject obj = new JSONObject();
+        obj.put("key", session.getAttribute("key"));
+        obj.put("new", passwordResetForm.getPassword1());
+
+        log.info("Connecting to sio for password reset...");
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        HttpEntity<String> request = new HttpEntity<>(obj.toString(), headers);
+        restTemplate.setErrorHandler(new MyResponseErrorHandler());
+        ResponseEntity response = null;
+        try {
+            response = restTemplate.exchange(properties.getPasswordResetURI(), HttpMethod.PUT, request, String.class);
+        } catch (RestClientException e) {
+            log.warn("Error connecting to sio for password reset! {}", e);
+            passwordResetForm.setErrMsg("Cannot connect to server! Please try again later.");
+            return FORGET_PSWD_NEW_PSWD_PAGE;
+        }
+
+        if (RestUtil.isError(response.getStatusCode())) {
+            Map<ExceptionState, String> exceptionMessageMap = new HashMap<>();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Convert this Map to an EnumMap. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1640)